### PR TITLE
#184581172 User Sign Up

### PR DESCRIPTION
**Pull request by 1MuhireDavid**
#### What does this PR do?
This PR allows enable user to be able to sign up.

#### Description of Task to be completed?
Have the following endpoint working
`POST /user/create`

#### How should this be manually tested?
- Clone this repo git clone https://github.com/atlp-rwanda/e-comm-team-amigos-bn
- Switch your working directory cd e-comm-team-amigos-bn
- Create an .env file and update it with .env-example
- Install all the dependencies by running npm install
- Then run the project by typing npm start
- Using Postman test **user/create** endpoint  
 by writing POST http://localhost:3000/user/create
_key_:Context-Type value: application/json

  

#### Any background context you want to provide?
- Have Node Js installed in your local machine
- Have Postgress and Sequelize installed on local machine
- Have Postman installed on your local machine

#### What are the relevant pivotal tracker/Trello stories?

Story ID: [ #184581172](https://www.pivotaltracker.com/n/projects/2632829/stories/184629360)

#### documented endponts on swager ui
Screenshot (161)

#### Passed Test
Screenshot (160)

### CODE REVIEW

**Review by AbdulKhaliq59 — APPROVED** [review]

**Review by Emile-SHUMBUSHO — APPROVED** [review]

**Review by GasanaJr — APPROVED** [review]

**Review by AbdulKhaliq59 — APPROVED** [review]

File: .env.example [hunk 1]
```diff
@@ -12,3 +12,7 @@ PORT=4000
 COVERALLS_REPO_TOKEN=covernal_token
 # PRODUCTION DB: REMOTE
 DATABASE_URL_PROD=production_url
+
+# SECRET TO BE USED IN THE HASHING PASSWORD AND IN TOKENS
+ACCESS_TOKEN_SECRET=766eb85b713df63f7285459288b790c8e9e6ab43229de9dca33fe08c1c5cc30961a83456c59c9aeefafc1db3bb8d27f6964572a456fec35d2d353874409ccc78
+REFRESH_TOKEN_SECRET=a00a248ce550b46b66b08cf5ed597a0b14b1877c02714110bfe402ded54a4a3e76b426c8da9f65726c33dbe83f736ed9d905bafd4a7e942bb0e271bd499a7c6b
```
Comment (nshutijonathan) [inline .env.example]: I hope these tokens are not valid ones, otherwise, we exposed them to the public.

File: src/middleware/authValidation.js [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+        return value;
+      }),
+      password: Joi.string().required().min(8),
+  });
+
+  try {
+    await schema.validateAsync(req.body);
+    next();
+  } catch (error) {
+    console.log(error);
```
Comment (nshutijonathan) [inline src/middleware/authValidation.js]: console.log is for debugging purpose ,remove it here ,It should not be pushed

**Review by nshutijonathan — CHANGES_REQUESTED** [review]
- Please you have more than two commits and that is not acceptable 
- Can you attach the screenshot showing passing tests?
- Can you also attach a screenshot showing swagger in action /speaking of user sign up 
- Please also update your readme file and inform non-technical users how your JSON values get passed. check readme of the following repo `https://github.com/atlp-rwanda/e-comm-team-brogrammers-bn`

File: package.json [hunk 3]
```diff
@@ -32,19 +32,25 @@
     "@babel/plugin-transform-runtime": "^7.21.0",
     "@babel/preset-env": "^7.20.2",
     "@babel/register": "^7.21.0",
+    "babel": "^6.23.0",
```
Comment (Habinezajanvier) [inline package.json]: - Remove babel in favor to babel/cli
- Remove also body-parser in favor to express.json()

File: package.json [hunk 4]
```diff
@@ -57,6 +63,7 @@
     "mocha-lcov-reporter": "^1.3.0",
     "nodemon": "^2.0.21",
     "nyc": "^15.1.0",
+    "rimraf": "^4.1.2",
```
Comment (Habinezajanvier) [inline package.json]: `rimraf` is installed twice

File: src/app.js [hunk 5]
```diff
@@ -1,26 +1,28 @@
...
 app.use(express.json());
 
-if (process.env.NODE_ENV === 'development') {
-  app.use(logger('dev'));
+if (process.env.NODE_ENV === "development") {
+  app.use(logger("dev"));
 }
 
-app.use('/user', userRoute);
+app.use(bodyParser.json());
```
Comment (Habinezajanvier) [inline src/app.js]: Remain only with `express.json()`

File: src/controllers/authentication.js [hunk 6]
```diff
@@ -0,0 +1,34 @@
+import { v4 as uuidv4 } from "uuid";
+import models from "../database/models";
+import bcrypt from "bcrypt";
+import tokenGenerator from "../helpers/generateToken";
+
+const createUser = async (req, res) => {
+  let user = await models.User.findOne({ where: { email: req.body.email } });
+  if (user) return res.status(400).send("Email address already in use (email)");
```
Comment (Habinezajanvier) [inline src/controllers/authentication.js]: Please don't send plain string as response

**Review by Habinezajanvier — CHANGES_REQUESTED** [review]

**Comment by ericndungutse** [comment]
This check is repeared twice. In auth controller and in validation. **let user = await models.User.findOne({ where: { email: req.body.email } }); 
   if (user) return res.status(400).send("Email address already in use (email)");**

**Comment by Emile-SHUMBUSHO** [comment]
> This check is repeared twice. In auth controller and in validation. **let user = await models.User.findOne({ where: { email: req.body.email } });   if (user) return res.status(400).send("Email address already in use (email)");**

**Review by Habinezajanvier — APPROVED** [review]
👍

**Pull request merged**